shared-libs/lineage/src/lineage.js

```diff
@@ -0,0 +1,403 @@
+  });
+
+  return doc;
+};
+
+var fillContactsInDocs = function(docs, contacts) {
+  if (!contacts || !contacts.length) {
+    return;
+  }
+  contacts.forEach(function(contactDoc) {
+    docs.forEach(function(doc) {
+      var id = doc && doc.contact && doc.contact._id;
+      if (id === contactDoc._id) {
+        doc.contact = contactDoc;
+      }
+    });
+  });
+};
+
+var fetchContacts = function(lineage) {
+  var contactIds = _.uniq(
+    lineage
+      .map(function(doc) {
+        return doc && doc.contact && doc.contact._id;
+      })
+      .filter(function(id) {
+        return !!id;
+      })
+  );
+
+  // Only fetch docs that are new to us
+  var lineageContacts = [];
+  var contactsToFetch = [];
+  contactIds.forEach(function(id) {
+    var contact = lineage.find(function(doc) {
+      return doc && doc._id === id;
+    });
+    if (contact) {
+      lineageContacts.push(contact);
+    } else {
+      contactsToFetch.push(id);
+    }
+  });
+
+  return fetchDocs(contactsToFetch)
+    .then(function(fetchedContacts) {
+      return lineageContacts.concat(fetchedContacts);
+    });
+};
+
+var mergeLineagesIntoDoc = function(lineage, contacts, patientLineage) {
+  patientLineage = patientLineage || [];
+  var lineages = lineage.concat(patientLineage);
+  fillContactsInDocs(lineages, contacts);
+
+  var doc = lineage.shift();
+  fillParentsInDocs(doc, lineage);
+
+  if (patientLineage.length) {
+    var patientDoc = patientLineage.shift();
+    fillParentsInDocs(patientDoc, patientLineage);
+    doc.patient = patientDoc;
+  }
+
+  return doc;
+};
+
+var findPatientId = function(doc) {
+  return (
+    doc.type === 'data_record' &&
+    ((doc.fields && doc.fields.patient_id) || doc.patient_id)
+  );
+};
+
+var fetchPatientLineage = function(record) {
+  var patientId = findPatientId(record);
+
+  if (!patientId) {
+    return Promise.resolve([]);
+  }
+
+  return contactUuidByPatientId(patientId)
+    .then(function(uuid) {
+      return fetchLineageById(uuid);
+    });
+};
+
+var contactUuidByPatientId = function(patientId) {
+  return DB.query('medic-client/contacts_by_reference', {
+    key: [ 'shortcode', patientId ]
+  }).then(function(results) {
+    if (results.rows.length > 1) {
+      console.warn('More than one patient person document for shortcode ' + patientId);
+    }
+
+    return results.rows[0] && results.rows[0].id;
+  });
+};
+
+var fetchLineageById = function(id) {
+  var options = {
+    startkey: [id],
+    endkey: [id, {}],
+    include_docs: true
+  };
+  return DB.query('medic-client/docs_by_id_lineage', options)
+    .then(function(result) {
+      return result.rows.map(function(row) {
+        return row.doc;
+      });
+    });
+};
+
+var fetchDoc = function(id) {
+  return DB.get(id)
+    .catch(function(err) {
+      if (err.status === 404) {
+        err.statusCode = 404;
+      }
+      throw err;
+    });
+};
+
+var fetchHydratedDoc = function(id, callback) {
```

Comment:
Why does this one use callbacks when all the other functions are promise based? Can we make them consistent?

Reply:
It support both.